chain/consensus/mir/state_manager.go

```diff
@@ -0,0 +1,263 @@
+	for i := 0; i < ConfigOffset+2; i++ {
+		memberships[i] = initialMembership
+	}
+	sm := StateManager{
+		ChainNotify:          make(chan []Messages),
+		MirManager:           m,
+		memberships:          memberships,
+		currentEpoch:         0,
+		reconfigurationVotes: make(map[string]uint64),
+	}
+	return &sm
+}
+
+func (sm *StateManager) ApplyEvents(eventsIn *events.EventList) (*events.EventList, error) {
+	return modules.ApplyEventsSequentially(eventsIn, sm.ApplyEvent)
+}
+
+func (sm *StateManager) ApplyEvent(event *eventpb.Event) (*events.EventList, error) {
+	switch e := event.Type.(type) {
+	case *eventpb.Event_Init:
+		return events.EmptyList(), nil
+	case *eventpb.Event_Deliver:
+		return sm.applyBatch(e.Deliver.Batch)
+	case *eventpb.Event_NewEpoch:
+		return sm.applyNewEpoch(e.NewEpoch)
+	case *eventpb.Event_AppSnapshotRequest:
+		return sm.applySnapshotRequest(e.AppSnapshotRequest)
+	case *eventpb.Event_AppRestoreState:
+		return sm.applyRestoreState(e.AppRestoreState.Snapshot)
+	default:
+		return nil, fmt.Errorf("unexpected type of App event: %T", event.Type)
+	}
+}
+
+// applyBatch applies a batch of requests to the state of the application.
+func (sm *StateManager) applyBatch(in *requestpb.Batch) (*events.EventList, error) {
+	var msgs []Messages
+
+	for _, req := range in.Requests {
+		switch req.Req.Type {
+		case TransportType:
+			msgs = append(msgs, req.Req.Data)
+		case ReconfigurationType:
+			err := sm.applyConfigMsg(req.Req)
+			if err != nil {
+				return events.EmptyList(), err
+			}
+		}
+	}
+
+	// Send messages to the Eudico node.
+	sm.ChainNotify <- msgs
+
+	return events.EmptyList(), nil
+}
+
+func (sm *StateManager) applyConfigMsg(in *requestpb.Request) error {
+	fmt.Printf("%v - applyConfigMsg, memb len - %d\n", sm.MirManager.MirID, len(sm.memberships))
+	newValSet := &hierarchical.ValidatorSet{}
+	if err := newValSet.UnmarshalCBOR(bytes.NewReader(in.Data)); err != nil {
+		return err
+	}
+	voted, err := sm.UpdateAndCheckVotes(newValSet)
+	if err != nil {
+		return err
+	}
+	if voted {
+		err = sm.UpdateNextMembership(newValSet)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// applyNewEpoch is applied ?
+//
+func (sm *StateManager) applyNewEpoch(newEpoch *eventpb.NewEpoch) (*events.EventList, error) {
+	sm.membershipLock.Lock()
+	defer sm.membershipLock.Unlock()
+
+	// Sanity check.
+	if t.EpochNr(newEpoch.EpochNr) != sm.currentEpoch+1 {
+		return nil, fmt.Errorf("expected next epoch to be %d, got %d", sm.currentEpoch+1, newEpoch.EpochNr)
+	}
+
+	// The base membership is the last one membership.
+	newMembership := sm.memberships[newEpoch.EpochNr+ConfigOffset]
+
+	// Append a new membership data structure to be modified throughout the new epoch.
+	sm.memberships = append(sm.memberships, newMembership)
+
+	// Update current epoch number.
+	sm.currentEpoch = t.EpochNr(newEpoch.EpochNr)
+
+	err := sm.MirManager.ReconfigureMirNode(context.TODO(), newMembership)
```

Comment:
This doesn't seem like the nicest way to communicate with the MirManager. The context for the reconfiguration is not available here. To use the miner's context, you can communicate the new membership to a channel (similarly as you now do with the transaction batches), and the main MirManager loop would call the `ReconfigureMirNode` method (which currently seems to consists only of connecting to the new nodes). It seems cleaner that way.

Reply:
Fixed - https://github.com/filecoin-project/eudico/pull/234/commits/5d32acbac73233852da1d5726be33565778736a8